feat: add the new Iterable plugin.

pauloouriques opened the pull request:
- add new iterable plugin as a destination;
- add simple api test call to test itarable api integration.

pauloouriques committed:
feat: add the new Iterable plugin.

- add new iterable plugin as a destination;
- add simple api test call to test itarable api integration.

grouparoo-bot committed:
Updating things automatically

bleonard commented on plugins/@grouparoo/iterable/README.md:
```diff
@@ -0,0 +1,5 @@
+
+## Attribution
+
+The Iterable icon used in this plugin is from http://getlogo.net/
```
ideally the whole url for attribution of where downloaded.

bleonard commented on plugins/@grouparoo/iterable/__tests__/.env.example:
```diff
@@ -0,0 +1 @@
+HUBSPOT_API_KEY = "a31ef454-1835-4b88-8d3a-eefc0b688ded"
```
will want to switch to `ITERABLE_API_KEY` when do tests.

bleonard commented on plugins/@grouparoo/iterable/src/initializers/plugin.ts:
```diff
@@ -0,0 +1,72 @@
+      name: packageJSON.name,
+      icon: "/public/@grouparoo/iterable/iterable.svg",
+      templates: [
+        new AppTemplate("hubspot", [
```
`hubspot` -> `iterable` updates throughout this file.

bleonard commented on plugins/@grouparoo/iterable/src/initializers/plugin.ts:
```diff
@@ -0,0 +1,72 @@
+          name: "iterable-export",
+          direction: "export",
+          description:
+            "Export Profiles to Iterable and add them to Contact Lists.",
```
"Export profiles to Iterable as Users and put them in static Lists"

bleonard commented on plugins/@grouparoo/iterable/src/lib/connect.ts:
```diff
@@ -0,0 +1,5 @@
+import { SimpleAppOptions } from "@grouparoo/core";
+
+export async function connect(appOptions: SimpleAppOptions) {
+  return require('node-iterable-api').create(appOptions.apiKey);
```
Let's `import Iterable from 'node-iterable-api'` (ideally) or `cosnt Iterable = require('node-iterable-api')` (if necessary) above and then do `Iterable.create()` here.

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/destinationOptions.ts:
```diff
@@ -0,0 +1,5 @@
+import { DestinationOptionsMethod } from "@grouparoo/core";
+
+export const destinationOptions: DestinationOptionsMethod = async () => {
+  return {};
```
will be empty (as-is)

bleonard commented on plugins/@grouparoo/iterable/src/lib/test.ts:
```diff
@@ -0,0 +1,9 @@
+export const test: TestPluginMethod = async ({ appOptions }) => {
+  const client = await connect(appOptions);
+  const success = await client.lists.get();
+  const message = success ? 'Connected to Iterable api.' : null;
```
What is success here? An array of lists?
We try to have a message something like.

```
  const response = await client.lists.get();
  const success = !!response.lists;
  const message = success ? 'Found ${response.lists.length} lists' : `Error getting lists`;
```

bleonard reviewed (commented):
Looking good!

Next step: destination mapping options using the fields

bleonard commented:
For Issue: https://github.com/grouparoo/grouparoo/issues/651

pauloouriques committed:
chore: add Iterable exportProfile.

- Add exportProfile jest tests (so far, only includes user creation);
- Add the destination mapping options tests.

bleonard commented on plugins/@grouparoo/hubspot/__tests__/integration/hubspot-export.ts:
```diff
@@ -1,4 +1,4 @@
-import path from "path";
+import * as path from "path";
```
Is this some autoformatting tool you have?

bleonard commented on plugins/@grouparoo/iterable/README.md:
```diff
@@ -0,0 +1,5 @@
+
+## Attribution
+
+The Iterable icon used in this plugin is from http://getlogo.net/iterable-inc-logo-vector-svg
```
I think maybe these are usually square. Can you find one of those. `png` is fine.

bleonard commented on plugins/@grouparoo/iterable/__tests__/.env.example:
```diff
@@ -0,0 +1 @@
+ITERABLE_API_KEY = ""
```
i usually put an example in here (not the real one of course) but it helps others to know what it should look like.

bleonard commented on plugins/@grouparoo/iterable/__tests__/export/destination-mapping-options.ts:
```diff
@@ -0,0 +1,75 @@
+// const newNock = false;
+// require("./../fixtures/destination-mapping-options");
+// or these to make it true
+const newNock = true;
```
will comment out for CI

bleonard commented on plugins/@grouparoo/iterable/__tests__/export/destination-mapping-options.ts:
```diff
@@ -0,0 +1,75 @@
+  test("can fetch user fields", async () => {
+    const client = await connect(appOptions);
+    const fields = await getUserFields(client);
+    const text = fields.find((f) => f.key === "phoneNumber");
```
these variable names don't look quite right. I would think this would would be `phoneNumber` too.

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/destinationMappingOptions.ts:
```diff
@@ -0,0 +1,75 @@
+      const type: DestinationMappingOptionsResponseTypes = mapTypesFromIterableToGrouparoo(
+        value
+      );
+      out.push({ key, type, important: true });
```
only push `if (type)`

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/destinationMappingOptions.ts:
```diff
@@ -0,0 +1,75 @@
+  };
+  const grouparooType = map[iterableType];
+  if (!grouparooType) {
+    throw `Unknown iterable type: ${iterableType}`;
```
we found it's better to `log` and allow this function to return `undefined` or `null`

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/destinationMappingOptions.ts:
```diff
@@ -0,0 +1,75 @@
+  key: string;
+  type: DestinationMappingOptionsResponseTypes;
+}> => {
+  return [{ key: "email", type: "string" }];
```
there is a type called "email" that this should use.

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/destinationMappingOptions.ts:
```diff
@@ -0,0 +1,75 @@
+};
+
+const mapTypesFromIterableToGrouparoo = (iterableType) => {
+  const map = {
```
if there is a special "known" ones that should be a `phoneNumber`, can can pass the key in here and catch it earlier.

bleonard replied:
It's usually a case statement at the top then the falls through to this mapping in the default type. You'd pass in the field name. SO if it's `phoneNumber` jsut said it's that type.

bleonard commented on plugins/@grouparoo/zendesk/__tests__/integration/destination-mapping-options.ts:
```diff
@@ -1,13 +1,10 @@
-import path from "path";
```
can you undo the changes to this file (and all the zendesk ones)? I feel like you have something removing unused things or something.
that gets weird when we do this comment in and out the nock thing.

bleonard commented on plugins/@grouparoo/iterable/__tests__/export/destination-mapping-options.ts:
```diff
@@ -0,0 +1,75 @@
+
+    const requiredFieldEmail = required.find((f) => f.key === "email");
+    expect(requiredFieldEmail.key).toBe("email");
+    expect(requiredFieldEmail.type).toBe("string");
```
should be "email"

bleonard commented on plugins/@grouparoo/iterable/__tests__/utils/shared.ts:
```diff
@@ -0,0 +1,8 @@
+const delayTime = 3 * 1000;
+
+export async function indexContacts() {
+  // search index takes a while to catch up after important change or add
```
it's ok to put this in the one test file.
Also only sleep `if newNock`

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/exportProfile.ts:
```diff
@@ -0,0 +1,77 @@
+  },
+}) => {
+  // if we received no mapped data... just exit
+  if (Object.keys(newProfileProperties).length === 0) {
```
Is this in others? That's weird. There still might be something to remove. I guess there's also no email?

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/exportProfile.ts:
```diff
@@ -0,0 +1,77 @@
+  try {
+    let user;
+    try {
+      user = await client.users.get({ email });
```
I thin it's possible you won't have to do this because it uses the email to upsert and add/remove from lists.

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/exportProfile.ts:
```diff
@@ -0,0 +1,77 @@
+      }
+    }
+    if (toDelete) {
+      if (user) {
```
if we don't have the user, can use just `deleteByEmail` and be sure to have a test where the user doesn't actually exist, so we can handle that.

bleonard reviewed (commented):
Looking good!
Made some comments and hints for the future.

pauloouriques committed:
feat: lists handing feature and tests.

pauloouriques committed:
chore: destination mapping tests fix on newNock configs.

bleonard commented on apps/staging-public/package.json:
```diff
@@ -31,6 +31,9 @@
     "@grouparoo/ui": "^0.2.0-alpha.15",
     "@grouparoo/zendesk": "^0.2.0-alpha.15"
   },
+  "devDependencies": {
```
this should be in the regular list (not dev). i may have said "dev" incorrectly before.

bleonard commented on plugins/@grouparoo/iterable/__tests__/export/export-profile.ts:
```diff
@@ -0,0 +1,387 @@
+
+async function deleteLists(suppressErrors) {
+  try {
+    for (const groupToDelete in ["List One", "List Two", "List Three"]) {
```
let's make instance variables out of these above to make sure they are the same as in use.

bleonard commented on plugins/@grouparoo/iterable/__tests__/export/export-profile.ts:
```diff
@@ -0,0 +1,387 @@
+  });
+  beforeAll(async () => {
+    apiClient = await connect(appOptions);
+    await deleteUsers(false);
```
other tests have standardized on a `cleanUp` function that does both of these things (users and lists)

bleonard commented on plugins/@grouparoo/iterable/__tests__/export/export-profile.ts:
```diff
@@ -0,0 +1,387 @@
+    user = await getUser(email);
+
+    expect(user).not.toBe(null);
+    expect(user.email).toBe(email);
```
check name

bleonard commented on plugins/@grouparoo/iterable/__tests__/export/export-profile.ts:
```diff
@@ -0,0 +1,387 @@
+      newProfileProperties: {
+        email,
+        userId,
+        name: newName,
```
i deally, we can test all the types here (iterable types: long, integer, etc)

bleonard commented on plugins/@grouparoo/iterable/__tests__/export/export-profile.ts:
```diff
@@ -0,0 +1,387 @@
+        userId,
+      },
+      oldGroups: [],
+      newGroups: ["List One", "List Two"],
```
use global instance variable

bleonard commented on plugins/@grouparoo/iterable/__tests__/export/export-profile.ts:
```diff
@@ -0,0 +1,387 @@
+    const user = await getUser(email);
+    expect(user.dataFields.emailListIds.length).toBe(2);
+
+    const listOneId = await getListId("List One");
```
in other tests, we save this (`listOneId`) globally and use it the rest of the time so don't look up each time.

bleonard commented on plugins/@grouparoo/iterable/__tests__/export/export-profile.ts:
```diff
@@ -0,0 +1,387 @@
+    expect(listTwoId).not.toBe(null);
+    expect(listOneThree).not.toBe(null);
+
+    expect(user.dataFields.emailListIds).toContain(listTwoId);
```
Is this one supposed to be checking for listOne (the already subscribed list that wasn't passed in)

pauloouriques replied:
yes

bleonard commented on plugins/@grouparoo/iterable/__tests__/export/export-profile.ts:
```diff
@@ -0,0 +1,387 @@
+    });
+    await indexContacts(newNock);
+
+    const user = await getUser(alternativeEmail);
```
check that the other email is not there.

bleonard commented on plugins/@grouparoo/iterable/src/initializers/plugin.ts:
```diff
@@ -0,0 +1,72 @@
+              key: "apiKey",
+              displayName: "Iterable API Key",
+              required: true,
+              description: "Iterable apikey (api) key.",
```
Maybe we could say where to find it here.

bleonard commented on plugins/@grouparoo/iterable/__tests__/export/export-profile.ts:
```diff
@@ -0,0 +1,387 @@
+        name: newName,
+        phoneNumber: newPhoneNumber,
+      },
+      oldGroups: [],
```
This plugin also supports random un-time properties. I'm not sure if we should literally make up a random string (not currently in getFields) and send it or not and make sure it works.

bleonard commented on plugins/@grouparoo/iterable/package.json:
```diff
@@ -0,0 +1,55 @@
+  "dependencies": {
+    "@grouparoo/app-templates": "^0.2.0-alpha.15",
+    "axios": "0.21.1",
+    "hubspot-api": "2.15.3",
```
can remove this hubspot now and maybe axios too.
are use using app-templates?

bleonard commented on plugins/@grouparoo/iterable/public/templates/app/{{{id}}}.js.template:
```diff
@@ -0,0 +1,13 @@
+      name: {{{id}}},
+      type: "{{{__pluginName}}}",
+      options: {
+        hapikey: "abc123", // Iterable hapikey (api) key
```
i think this name is wrong (with the h)

bleonard commented on plugins/@grouparoo/iterable/public/templates/destination/{{{id}}}.js.template:
```diff
@@ -0,0 +1,28 @@
+      name: {{{id}}},
+      class: "destination",
+      type: "{{{__pluginName}}}-export",
+      appId: "hubspotApp", // The id of the App this Source uses
```
todo

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/destinationMappingOptions.ts:
```diff
@@ -0,0 +1,77 @@
+  return {
+    labels: {
+      property: {
+        singular: "Iterable Contact Property",
```
Does iterable UI call them property or fields?

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/exportProfile.ts:
```diff
@@ -0,0 +1,98 @@
+  delete oldProfileProperties.email;
+  delete newProfileProperties.email;
+
+  try {
```
maybe make a submethod (like intercom) for this whole thing and try/catch in the root.

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/exportProfile.ts:
```diff
@@ -0,0 +1,98 @@
+
+      if (currentEmail && currentEmail !== email) {
+        const emailPayload = { currentEmail, newEmail: email };
+        if (newProfileProperties.userId) {
```
it's interesting that this could also be the `old` userId.

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/exportProfile.ts:
```diff
@@ -0,0 +1,98 @@
+          await removeFromList(client, appGuid, appOptions, email, group);
+      }
+
+      if (currentEmail && currentEmail !== email) {
```
doesn't this email change have to happen _before_ using the email to add/remove from groups?

bleonard replied:
let's make sure we have tests with to cover that case and then fix. I'm interested to know if this method will have to wait or not like `indexContacts` to make it work.

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/exportProfile.ts:
```diff
@@ -0,0 +1,98 @@
+  }
+
+  // Remove email from profile properties to avoid duplicates.
+  delete oldProfileProperties.email;
```
usually we don't change these inputs, but skip "email" in payload building if that is why you did it.

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/exportProfile.ts:
```diff
@@ -0,0 +1,98 @@
+
+  try {
+    if (toDelete) {
+      await client.users.delete(email);
```
it's ok to return early here to prevent the indentation for the rest of the method.

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/exportProfile.ts:
```diff
@@ -0,0 +1,98 @@
+
+      const dataFields = Object.assign(
+        {},
+        newProfileProperties,
```
it's a minor thing, but I would suggest the other other order to make sure that `newProperfileProperties` always "wins"
```
const dataFields = Object.assign(
        {},
        deletePropertiesPayload
        newProfileProperties
)
```

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/listMethods.ts:
```diff
@@ -0,0 +1,85 @@
+  groupName
+) {
+  const listId = await getListId(client, appGuid, appOptions, groupName);
+  await client.lists.unsubscribe({ listId, subscribers: [{ email }] });
```
Do we ahve a test if the user i not in this list? We might have to catch it.

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/exportProfile.ts:
```diff
@@ -0,0 +1,98 @@
+
+      // add to lists
+      for (const groupToAdd of newGroups) {
+        await addToList(client, appGuid, appOptions, email, groupToAdd);
```
does the response from `client.users.update(payload);` above have the listIds the user is in? You could use that here to only add if they aren't in the list (and remove if only if they are in the list).
This would save API calls.

pauloouriques replied:
nope, the response shows only the status.

bleonard replied:
It's interesting that the `getUser` (from tests) call seems to have that data.
There's some number (2+) of lists where making that call is the more efficient thing. But then there is also the timing issue of the user showing up. Not great.

Maybe it's actually better to put that back at the beginning and use that data when making the list. It could also be combined with the change email situation to see that they are already there.


```
  let existingUser = await client.users.get(email);
  if (currentEmail && email !== currentEmail) {
    const oldUser = await client.users.get(oldEmail);
    if (oldUser) {
      ....
      await client.users.updateEmail(emailPayload);
      existingUser = oldUser;
      existingUser.email = email;
    }
  }
```
This makes the switchover much smarter at the expense of these 2 calls. and one of them makes the update lists case possible much more efficent.

Then you'll have `existingUser` lower down to use if it's not `null`. use `existingUser.dataFields.emailListIds` to make add/remove smarter.

bleonard reviewed (commented):
On the right track!
Lots of comments, but nothing major.

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/listMethods.ts:
```diff
@@ -0,0 +1,85 @@
+): Promise<string> {
+  groupName = (groupName || "").toString().trim();
+  if (groupName.length === 0) {
+    return `Iterable empty groupName`;
```
not sure about this.

bleonard replied:
Not sure if the other plugins handle this. it's a good idea. --- let's try something more user facing like `"Grouparoo Default"`

pauloouriques committed:
chore: review comments changes.

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/exportProfile.ts:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  if (toDelete) {
+    await client.users.delete(email);
```
note, it could also be the old user to delete.

pauloouriques replied:
can you explain better?

bleonard replied:
The user that needs to be deleted could be `currentEmail` from `oldProperties`. If you went with the code I referenced in https://github.com/grouparoo/grouparoo/pull/1204#discussion_r566486570, then you'd maybe add deleting to that concept or do something like that to see who is really in there.

pauloouriques committed:
chore: last changes based on review.

- added lists check before subscribing/unsubscribing;
- check old email before delete a user.

bleonard commented on plugins/@grouparoo/hubspot/__tests__/utils/nockHelper.ts:
```diff
@@ -1,7 +1,7 @@
 import { SimpleAppOptions } from "@grouparoo/core";
```
One more file to put back as before.

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/exportProfile.ts:
```diff
@@ -0,0 +1,115 @@
+    return sendProfile(args);
+  } catch (error) {
+    if (error?.response?.status === 429) {
+      return { error, success: false, retryDelay: 1000 * 11 }; // the most common rate-limit error from hubspot is in 10-second intervals
```
is this how iterable really works?
just noting this still says "hubspot"

bleonard commented on pnpm-lock.yaml:
```diff
@@ -516,6 +519,42 @@ importers:
+    dependencies:
+      '@grouparoo/app-templates': link:../app-templates
+      axios: 0.21.1
+      hubspot-api: 2.15.3
```
weird this is still here.

bleonard reviewed (commented):
Quick hubspot check. will look more soon.

bleonard commented on plugins/@grouparoo/iterable/__tests__/export/destination-mapping-options.ts:
```diff
@@ -0,0 +1,75 @@
+    const client = await connect(appOptions);
+    const fields = await getUserFields(client);
+    const phoneNumber = fields.find((f) => f.key === "phoneNumber");
+    expect(phoneNumber.type).toBe("string");
```
seems likely this should be `phoneNumber` and the noted change in the mapping. It's ok to have default columns with known names like this.

Then maybe one more "first name" or something here to test for string.

bleonard commented on plugins/@grouparoo/iterable/__tests__/export/export-profile.ts:
```diff
@@ -0,0 +1,463 @@
+
+async function deleteLists(suppressErrors) {
+  try {
+    for (const groupToDelete in [listOne, listTwo, listThree]) {
```
probably `of` here. I think you are deleting "0" and "1" right now.

bleonard replied:
`listFour` in this list, too.

bleonard commented on plugins/@grouparoo/iterable/src/initializers/plugin.ts:
```diff
@@ -0,0 +1,73 @@
+            exportProfile,
+            destinationOptions,
+            destinationMappingOptions,
+            exportArrayProperties,
```
Note: next we'll want to look more into array properties and `getUserFields`

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/exportProfile.ts:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    await client.users.update(payload);
+    const currentUser = await getUser(client, email);
```
When this user is added, don't we have the issue of it taking N seconds. I think if you had a test that added a new user _and_ put them groups we might have a problem here in that it would be null. Or if you changed the email and added groups, it wouldn't find them here.

The code I suggested before had this above `if (currentEmail && currentEmail !== email)` and reassigned the `currentUser` to the changed one.
```
  let currentUser = await client.users.get(email);
  if (currentEmail && email !== currentEmail) {
    const oldUser = await client.users.get(oldEmail);
    if (oldUser) {
      ....
      await client.users.updateEmail(emailPayload);
      currentUser = oldUser;
      currentUser.email = email;
    }
  }
```

With that approach a null `currentUser` will be known to not be in any groups. and you won't have that `indexContacts` type issue.

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/listMethods.ts:
```diff
@@ -0,0 +1,100 @@
+  if (
+    !user.dataFields.emailListIds ||
+    !user.dataFields.emailListIds.includes(listId)
+  )
```
add `{` here to be clear it's all in the `if`

maybe even better to add semantics for clarity.
```
    const currentListIds = user?.dataFields?.emailListIds || [];
    if(currentListIds.includes(listId)) {
      return;
    }
    await client.lists.subscribe({
      listId,
      subscribers: [{ email: user.email }],
    });
```

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/listMethods.ts:
```diff
@@ -0,0 +1,100 @@
+  user,
+  groupName
+) {
+  const listId = await getListId(client, appGuid, appOptions, groupName);
```
same clarity idea as `add`, but in reverse.

bleonard reviewed (commented):
maybe still a few things about timing, especially on new user and change email.

bleonard commented on plugins/@grouparoo/iterable/__tests__/fixtures/destination-mapping-options.js:
```diff
@@ -0,0 +1,42 @@
+const nock = require('nock');
```
note: you'll want to prettier format these `nocks`

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/destinationMappingOptions.ts:
```diff
@@ -0,0 +1,77 @@
+        value
+      );
+      if (type) {
+        out.push({ key, type, important: true });
```
they probably aren't all `important`. We probably need a whitelist with these in it:
name
phoneNumber
userId
signupDate

I don't think we want the ones with `"."` to be `important`
Maybe that's a way to do it.

I think we also want to leave out these ones:
emailListIds
unsubscribedChannelIds
unsubscribedMessageTypeIds

pauloouriques committed:
chore: more changes based on the MR review.

pauloouriques committed:
fix: package.json missin a comma.

pauloouriques committed:
chore: add phone number type treatment to mapTypesFromIterableToGrouparoo.

pauloouriques committed:
fix: adjust test case to the mapping type change.

pauloouriques committed:
chore: remove user check during export-profile.

bleonard commented on plugins/@grouparoo/iterable/src/lib/export/exportProfile.ts:
```diff
@@ -69,32 +69,27 @@ export const sendProfile: ExportProfilePluginMethod = async ({
+      if (newProfileProperties.userId) {
+        emailPayload["currentUserId"] = newProfileProperties.userId;
       }
+      await client.users.updateEmail(emailPayload);
```
let's get a test to understand what happens here if the `old` email address is not in the system. Might have to catch.

pauloouriques committed:
chore: create new test cases to cover invalid props.

pauloouriques committed:
chore: bump up iterable version to 0.2.0-alpha.17.
